File: src/commands/autofixCommand.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open CommandUtils
open CommandSpec

(* This module implements the flow command `autofix insert-type` which inserts
   a type annotation in a file at a position. *)
module InsertType = struct
  let spec =
    Autofix_options.(
      let ambiguity_strategies_list = String.concat ", " @@ List.map fst ambiguity_strategies in
      {
        name = "insert type";
        doc = "[EXPERIMENTAL] Insert type information at file and position";
        usage =
          Printf.sprintf
            "Usage: %s autofix insert-type [OPTION]... [FILE] LINE COLUMN [END_LINE] [END_COLUMN]\n\ne.g. %s autofix insert-type foo.js 12 3\nor   %s autofix insert-type 12 3 < foo.js\n"
            exe_name
            exe_name
            exe_name;
        args =
          ArgSpec.(
            empty
            |> base_flags
            |> connect_and_json_flags
            |> root_flag
            |> strip_root_flag
            |> verbose_flags
            |> from_flag
            |> path_flag
            |> wait_for_recheck_flag
            |> flag
                 "--strict-location"
                 no_arg
                 ~doc:"Restrict the number of valid positions for each annotation"
            |> flag
                 "--strategy"
                 (required ~default:Generalize (enum ambiguity_strategies))
                 ~doc:
                   ( "Set how to resolve ambiguity in possible types ("
                   ^ ambiguity_strategies_list
                   ^ ")" )
            |> flag
                 "--in-place"
                 no_arg
                 ~doc:"Overwrite the input file or file specified by the path flag"
            |> flag "--expand-type-aliases" no_arg ~doc:"Replace type aliases with their bodies"
            |> flag
                 "--omit-typearg-defaults"
                 no_arg
                 ~doc:"Omit type arguments when defaults exist and match the provided type argument"
            |> anon "args" (required (list_of string)));
      })

  let handle_error ?(code = FlowExitStatus.Unknown_error) msg = FlowExitStatus.(exit ~msg code)

  let rec parse_args args : Loc.t =
    let parse_pos line col : Loc.position =
      let (line, column) =
        try convert_input_pos (int_of_string line, int_of_string col)
        with Failure _ -> handle_error "flow autofix insert-type: failed to parse position"
      in
      Loc.{ line; column }
    in
    match args with
    | [start_line; start_col; end_line; end_col] ->
      let start = parse_pos start_line start_col in
      let _end = parse_pos end_line end_col in
      Loc.{ source = None; start; _end }
    | [start_line; start_col] ->
      let start = parse_pos start_line start_col in
      Loc.{ source = None; start; _end = start }
    | file :: (([_; _] | [_; _; _; _]) as loc) ->
      let loc = parse_args loc in
      Loc.{ loc with source = Some (File_key.SourceFile (expand_path file)) }
    | [] -> handle_error "flow autofix insert-type: No position given"
    | _ -> handle_error "flow autofix insert-type: Invalid position given"

  let select_output_channel in_place path source_path =
    match (in_place, path, source_path) with
    | (false, _, _) -> stdout
    | (true, Some p, _)
    | (true, None, Some p) ->
      begin
        try open_out p
        with _ ->
          handle_error ~code:FlowExitStatus.Path_is_not_a_file
          @@ Printf.sprintf "failed to open output file: %s" p
      end
    | (true, None, None) ->
      handle_error "Flow: --in-place flag used without input file or explicit path"

  let handle_ok patch input in_place path source_path =
    match File_input.content_of_file_input input with
    | Ok content ->
      let out = select_output_channel in_place path source_path in
      output_string out @@ Replacement_printer.print patch content;
      close_out out
    | Error msg -> handle_error msg

  let main
      base_flags
      option_values
      json
      _pretty
      root_arg
      strip_root_arg
      verbose
      path
      wait_for_recheck
      location_is_strict
      ambiguity_strategy
      in_place
      expand_aliases
      omit_targ_defaults
      args
      () =
    let (Loc.{ source; _ } as target) = parse_args args in
    let source_path = Option.map ~f:File_key.to_string source in
    let input = get_file_from_filename_or_stdin ~cmd:spec.name path source_path in
    let root = get_the_root ~base_flags ~input root_arg in
    (* TODO Figure out how to implement root striping *)
    let _strip_root =
      if strip_root_arg then
        Some root
      else
        None
    in
    let flowconfig_name = base_flags.Base_flags.flowconfig_name in
    if (not json) && verbose <> None then
      prerr_endline "NOTE: --verbose writes to the server log file";
    let request =
      ServerProt.Request.INSERT_TYPE
        {
          input;
          target;
          verbose;
          location_is_strict;
          ambiguity_strategy;
          wait_for_recheck;
          expand_aliases;
          omit_targ_defaults;
        }
    in
    let result = connect_and_make_request flowconfig_name option_values root request in
    match result with
    | ServerProt.Response.INSERT_TYPE (Error err) -> handle_error err
    (* TODO implement a more useful set of error conditions *)
    | ServerProt.Response.INSERT_TYPE (Ok resp) -> handle_ok resp input in_place path source_path
    | _ -> handle_error "Flow: invalid server response"

  let command = CommandSpec.command spec main
end

module Exports = struct
  let spec =
    {
      name = "exports";
      doc = "[EXPERIMENTAL] automatically fix signature verification errors";
      usage = Printf.sprintf "Usage: %s autofix exports [OPTION]... [FILE]\n" exe_name;
      args =
        ArgSpec.(
          empty
          |> base_flags
          |> connect_and_json_flags
          |> root_flag
          |> strip_root_flag
          |> verbose_flags
          |> from_flag
          |> path_flag
          |> wait_for_recheck_flag
          |> flag
               "--in-place"
               no_arg
               ~doc:"Overwrite the input file or file specified by the path flag"
          |> flag "--force" no_arg ~doc:"Write the results even if errors are encountered"
          |> anon "file" (required string));
    }

  let handle_error ?(code = FlowExitStatus.Unknown_error) msg = FlowExitStatus.(exit ~msg code)

  let select_output_channel in_place path source_path =
    match (in_place, path, source_path) with
    | (false, _, _) -> stdout
    | (true, Some p, _)
    | (true, None, p) ->
      begin
        try open_out p
        with _ ->
          handle_error ~code:FlowExitStatus.Path_is_not_a_file
          @@ Printf.sprintf "failed to open output file: %s" p
      end

  let avg_error_size = 100

  let append_errors errors =
    Buffer.(
      let buff = create (avg_error_size * List.length errors) in
      List.fold_left (fun () -> add_string buff) () errors;
      contents buff)

  let handle_ok patch errors input in_place forced path source_path =
    let write_patch content =
      let out = select_output_channel in_place path source_path in
      output_string out @@ Replacement_printer.print patch content;
      close_out out
    in
    match (File_input.content_of_file_input input, errors, forced) with
    | (Ok content, [], _)
    | (Ok content, _, true) ->
      output_string stderr (append_errors errors);
      write_patch content
    | (Ok _, errors, false) -> handle_error (append_errors errors)
    | (Error msg, _, _) -> handle_error msg

  let main
      base_flags
      option_values
      json
      _pretty
      root_arg
      _strip_root_arg
      verbose
      path
      wait_for_recheck
      in_place
      forced
      source_path
      () =
    let source_path = expand_path source_path in
    let input = get_file_from_filename_or_stdin ~cmd:spec.name path (Some source_path) in
    let root = get_the_root ~base_flags ~input root_arg in
    let flowconfig_name = base_flags.Base_flags.flowconfig_name in
    if (not json) && verbose <> None then
      prerr_endline "NOTE: --verbose writes to the server log file";
    let request = ServerProt.Request.AUTOFIX_EXPORTS { input; verbose; wait_for_recheck } in
    let result = connect_and_make_request flowconfig_name option_values root request in
    match result with
    | ServerProt.Response.AUTOFIX_EXPORTS (Error err) -> handle_error err
    | ServerProt.Response.AUTOFIX_EXPORTS (Ok (patch, errors)) ->
      handle_ok patch errors input in_place forced path source_path
    | _ -> handle_error "Flow: invalid server response"

  let command = CommandSpec.command spec main
end

let command =
  let main (cmd, argv) () = CommandUtils.run_command cmd argv in
  let spec =
    {
      CommandSpec.name = "autofix";
      doc = "";
      usage =
        Printf.sprintf
          "Usage: %s autofix SUBCOMMAND [OPTIONS]...\n\nSUBCOMMANDS:\nsuggest: Provides type annotation suggestions for a given program\ninsert-type: Insert type information at file and position\nexports: Automatically fix signature verification errors\n"
          CommandUtils.exe_name;
      args =
        CommandSpec.ArgSpec.(
          empty
          |> anon
               "subcommand"
               (required
                  (command
                     [
                       ("suggest", SuggestCommand.command);
                       ("insert-type", InsertType.command);
                       ("exports", Exports.command);
                     ])));
    }
  in
  CommandSpec.command spec main

File: rust_port/crates/flow_cli/src/autofix_command.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::io::Write;

use flow_parser::file_key;
use flow_parser::file_key::FileKey;
use flow_parser::loc::Loc;
use flow_parser::loc::Position;
use flow_parser_utils_output::replacement_printer;
use flow_server_env::server_prot;
use flow_server_utils::file_input::FileInput;

use crate::command_spec;
use crate::command_spec::arg_spec;
use crate::command_utils;

// This module implements the flow command `autofix insert-type` which inserts
// a type annotation in a file at a position.
mod insert_type {
    use super::*;

    fn spec() -> command_spec::Spec {
        let exe_name = command_utils::exe_name();
        let spec = command_spec::Spec::new(
            "insert type",
            "[EXPERIMENTAL] Insert type information at file and position",
            command_spec::Visibility::Public,
            format!(
                "Usage: {exe_name} autofix insert-type [OPTION]... [FILE] LINE COLUMN [END_LINE] [END_COLUMN]\n\ne.g. {exe_name} autofix insert-type foo.js 12 3\nor   {exe_name} autofix insert-type 12 3 < foo.js\n"
            ),
        );
        let spec = command_utils::add_base_flags(spec);
        let spec = command_utils::add_connect_and_json_flags(spec);
        let spec = command_utils::add_root_flag(spec);
        let spec = command_utils::add_strip_root_flag(spec);
        let spec = command_utils::add_verbose_flags(spec);
        let spec = command_utils::add_from_flag(spec);
        let spec = command_utils::add_path_flag(spec);
        let spec = command_utils::add_wait_for_recheck_flag(spec);
        spec.flag(
            "--strict-location",
            &arg_spec::truthy(),
            "Restrict the number of valid positions for each annotation",
            None,
        )
        .flag(
            "--in-place",
            &arg_spec::truthy(),
            "Overwrite the input file or file specified by the path flag",
            None,
        )
        .flag(
            "--omit-typearg-defaults",
            &arg_spec::truthy(),
            "Omit type arguments when defaults exist and match the provided type argument",
            None,
        )
        .anon("args", &arg_spec::list_of(arg_spec::string()))
    }

    fn handle_error(code: flow_common_exit_status::FlowExitStatus, msg: &str) -> ! {
        eprintln!("{}", msg);
        flow_common_exit_status::exit(code);
    }

    // Returns (loc, source_path_opt). The source path is a raw absolute string;
    // File_key construction is deferred until after roots are initialized.
    fn parse_args(args: &[String]) -> (Loc, Option<String>) {
        let parse_pos = |line: &str, col: &str| -> Position {
            let line: i32 = line.parse().unwrap_or_else(|_| {
                handle_error(
                    flow_common_exit_status::FlowExitStatus::UnknownError,
                    "flow autofix insert-type: failed to parse position",
                )
            });
            let col: i32 = col.parse().unwrap_or_else(|_| {
                handle_error(
                    flow_common_exit_status::FlowExitStatus::UnknownError,
                    "flow autofix insert-type: failed to parse position",
                )
            });
            let (line, column) = command_utils::convert_input_pos(line, col);
            Position { line, column }
        };
        match args {
            [start_line, start_col, end_line, end_col] => {
                let start = parse_pos(start_line, start_col);
                let end = parse_pos(end_line, end_col);
                (
                    Loc {
                        source: None,
                        start,
                        end,
                    },
                    None,
                )
            }
            [start_line, start_col] => {
                let start = parse_pos(start_line, start_col);
                (
                    Loc {
                        source: None,
                        start,
                        end: start,
                    },
                    None,
                )
            }
            [file, rest @ ..] if rest.len() == 2 || rest.len() == 4 => {
                let (loc, _) = parse_args(rest);
                (loc, Some(command_utils::expand_path(file)))
            }
            [] => handle_error(
                flow_common_exit_status::FlowExitStatus::UnknownError,
                "flow autofix insert-type: No position given",
            ),
            _ => handle_error(
                flow_common_exit_status::FlowExitStatus::UnknownError,
                "flow autofix insert-type: Invalid position given",
            ),
        }
    }

    fn select_output_channel(
        in_place: bool,
        path: Option<&str>,
        source_path: Option<&str>,
    ) -> Box<dyn Write> {
        match (in_place, path, source_path) {
            (false, _, _) => Box::new(std::io::stdout()),
            (true, Some(p), _) | (true, None, Some(p)) => {
                Box::new(std::fs::File::create(p).unwrap_or_else(|_| {
                    handle_error(
                        flow_common_exit_status::FlowExitStatus::PathIsNotAFile,
                        &format!("failed to open output file: {}", p),
                    )
                }))
            }
            (true, None, None) => handle_error(
                flow_common_exit_status::FlowExitStatus::UnknownError,
                "Flow: --in-place flag used without input file or explicit path",
            ),
        }
    }

    fn handle_ok(
        patch: &replacement_printer::Patch,
        input: &FileInput,
        in_place: bool,
        path: Option<&str>,
        source_path: Option<&str>,
    ) {
        match input.content_of_file_input() {
            Ok(content) => {
                let mut out = select_output_channel(in_place, path, source_path);
                out.write_all(replacement_printer::print(patch, &content).as_bytes())
                    .unwrap();
                out.flush().unwrap();
            }
            Err(msg) => handle_error(flow_common_exit_status::FlowExitStatus::UnknownError, &msg),
        }
    }

    fn main(args: &arg_spec::Values) {
        let base_flags = command_utils::get_base_flags(args);
        let (connect_flags, json, _pretty) = command_utils::get_connect_and_json_flags(args);
        let root_arg =
            command_spec::get(args, "--root", &arg_spec::optional(arg_spec::string())).unwrap();
        let strip_root_arg = command_spec::get(args, "--strip-root", &arg_spec::truthy()).unwrap();
        let verbose = command_utils::verbose_flags(args);
        let path =
            command_spec::get(args, "--path", &arg_spec::optional(arg_spec::string())).unwrap();
        let wait_for_recheck = command_spec::get(
            args,
            "--wait-for-recheck",
            &arg_spec::optional(arg_spec::bool_flag()),
        )
        .unwrap();
        let location_is_strict =
            command_spec::get(args, "--strict-location", &arg_spec::truthy()).unwrap();
        let in_place = command_spec::get(args, "--in-place", &arg_spec::truthy()).unwrap();
        let omit_targ_defaults =
            command_spec::get(args, "--omit-typearg-defaults", &arg_spec::truthy()).unwrap();
        let raw_args = command_spec::get(args, "args", &arg_spec::list_of(arg_spec::string()))
            .unwrap()
            .unwrap_or_default();

        let (target, source_path) = parse_args(&raw_args);
        let input = command_utils::get_file_from_filename_or_stdin(
            &spec().name,
            source_path.as_deref(),
            path.as_deref(),
        );
        let root = command_utils::get_the_root(
            &base_flags.flowconfig_name,
            root_arg.as_deref(),
            Some(&input),
        );
        // Set root before constructing File_key so the suffix is root-relative,
        // matching the server's File_key values.
        file_key::set_project_root(&root.to_string_lossy());
        let mut target = target;
        match source_path.as_ref() {
            Some(abs) => {
                target.source = Some(FileKey::source_file_of_absolute(abs));
            }
            None => {}
        }
        // TODO Figure out how to implement root striping
        let _strip_root = if strip_root_arg {
            Some(root.clone())
        } else {
            None
        };
        let flowconfig_name = &base_flags.flowconfig_name;
        if !json && verbose.is_some() {
            eprintln!("NOTE: --verbose writes to the server log file");
        }
        let request = server_prot::request::Command::INSERT_TYPE {
            input: input.clone(),
            target,
            verbose,
            location_is_strict,
            wait_for_recheck,
            omit_targ_defaults,
        };
        let result = command_utils::connect_and_make_request(
            flowconfig_name,
            &connect_flags,
            &root,
            &request,
        );
        match result {
            server_prot::response::Response::INSERT_TYPE(Err(err)) => {
                handle_error(flow_common_exit_status::FlowExitStatus::UnknownError, &err)
            }
            // TODO implement a more useful set of error conditions
            server_prot::response::Response::INSERT_TYPE(Ok(resp)) => handle_ok(
                &resp,
                &input,
                in_place,
                path.as_deref(),
                source_path.as_deref(),
            ),
            _ => handle_error(
                flow_common_exit_status::FlowExitStatus::UnknownError,
                "Flow: invalid server response",
            ),
        }
    }

    pub(super) fn command() -> command_spec::Command {
        command_spec::command(spec(), main)
    }
}

mod missing_local_annot {
    use super::*;

    fn spec() -> command_spec::Spec {
        let exe_name = command_utils::exe_name();
        let spec = command_spec::Spec::new(
            "missing-local-annot",
            "[EXPERIMENTAL] automatically fix missing-local-annot errors",
            command_spec::Visibility::Public,
            format!("Usage: {exe_name} autofix missing-local-annot [OPTION]... [FILE]\n"),
        );
        let spec = command_utils::add_base_flags(spec);
        let spec = command_utils::add_connect_and_json_flags(spec);
        let spec = command_utils::add_root_flag(spec);
        let spec = command_utils::add_strip_root_flag(spec);
        let spec = command_utils::add_verbose_flags(spec);
        let spec = command_utils::add_from_flag(spec);
        let spec = command_utils::add_path_flag(spec);
        let spec = command_utils::add_wait_for_recheck_flag(spec);
        spec.flag(
            "--in-place",
            &arg_spec::truthy(),
            "Overwrite the input file or file specified by the path flag",
            None,
        )
        .anon("file", &arg_spec::required(None, arg_spec::string()))
    }

    fn handle_error(code: flow_common_exit_status::FlowExitStatus, msg: &str) -> ! {
        eprintln!("{}", msg);
        flow_common_exit_status::exit(code);
    }

    fn select_output_channel(
        in_place: bool,
        path: Option<&str>,
        source_path: &str,
    ) -> Box<dyn Write> {
        match (in_place, path) {
            (false, _) => Box::new(std::io::stdout()),
            (true, Some(p)) => Box::new(std::fs::File::create(p).unwrap_or_else(|_| {
                handle_error(
                    flow_common_exit_status::FlowExitStatus::PathIsNotAFile,
                    &format!("failed to open output file: {}", p),
                )
            })),
            (true, None) => Box::new(std::fs::File::create(source_path).unwrap_or_else(|_| {
                handle_error(
                    flow_common_exit_status::FlowExitStatus::PathIsNotAFile,
                    &format!("failed to open output file: {}", source_path),
                )
            })),
        }
    }

    #[allow(dead_code)]
    const AVG_ERROR_SIZE: usize = 100;

    fn handle_ok(
        patch: &replacement_printer::Patch,
        input: &FileInput,
        in_place: bool,
        path: Option<&str>,
        source_path: &str,
    ) {
        let write_patch = |content: &str| {
            let mut out = select_output_channel(in_place, path, source_path);
            out.write_all(replacement_printer::print(patch, content).as_bytes())
                .unwrap();
            out.flush().unwrap();
        };
        match input.content_of_file_input() {
            Ok(content) => write_patch(&content),
            Err(msg) => handle_error(flow_common_exit_status::FlowExitStatus::UnknownError, &msg),
        }
    }

    fn main(args: &arg_spec::Values) {
        let base_flags = command_utils::get_base_flags(args);
        let (connect_flags, json, _pretty) = command_utils::get_connect_and_json_flags(args);
        let root_arg =
            command_spec::get(args, "--root", &arg_spec::optional(arg_spec::string())).unwrap();
        let _strip_root_arg = command_spec::get(args, "--strip-root", &arg_spec::truthy()).unwrap();
        let verbose = command_utils::verbose_flags(args);
        let path =
            command_spec::get(args, "--path", &arg_spec::optional(arg_spec::string())).unwrap();
        let wait_for_recheck = command_spec::get(
            args,
            "--wait-for-recheck",
            &arg_spec::optional(arg_spec::bool_flag()),
        )
        .unwrap();
        let in_place = command_spec::get(args, "--in-place", &arg_spec::truthy()).unwrap();
        let file =
            command_spec::get(args, "file", &arg_spec::required(None, arg_spec::string())).unwrap();

        let source_path = command_utils::expand_path(&file);
        let input = command_utils::get_file_from_filename_or_stdin(
            &spec().name,
            Some(source_path.as_str()),
            path.as_deref(),
        );
        let root = command_utils::get_the_root(
            &base_flags.flowconfig_name,
            root_arg.as_deref(),
            Some(&input),
        );
        let flowconfig_name = &base_flags.flowconfig_name;
        if !json && verbose.is_some() {
            eprintln!("NOTE: --verbose writes to the server log file");
        }
        let request = server_prot::request::Command::AUTOFIX_MISSING_LOCAL_ANNOT {
            input: input.clone(),
            verbose,
            wait_for_recheck,
        };
        let result = command_utils::connect_and_make_request(
            flowconfig_name,
            &connect_flags,
            &root,
            &request,
        );
        match result {
            server_prot::response::Response::AUTOFIX_MISSING_LOCAL_ANNOT(Err(err)) => {
                handle_error(flow_common_exit_status::FlowExitStatus::UnknownError, &err)
            }
            server_prot::response::Response::AUTOFIX_MISSING_LOCAL_ANNOT(Ok(patch)) => {
                handle_ok(&patch, &input, in_place, path.as_deref(), &source_path)
            }
            _ => handle_error(
                flow_common_exit_status::FlowExitStatus::UnknownError,
                "Flow: invalid server response",
            ),
        }
    }

    pub(super) fn command() -> command_spec::Command {
        command_spec::command(spec(), main)
    }
}

mod exports {
    use super::*;

    fn spec() -> command_spec::Spec {
        let exe_name = command_utils::exe_name();
        let spec = command_spec::Spec::new(
            "exports",
            "[EXPERIMENTAL] automatically fix signature verification errors",
            command_spec::Visibility::Public,
            format!("Usage: {exe_name} autofix exports [OPTION]... [FILE]\n"),
        );
        let spec = command_utils::add_base_flags(spec);
        let spec = command_utils::add_connect_and_json_flags(spec);
        let spec = command_utils::add_root_flag(spec);
        let spec = command_utils::add_strip_root_flag(spec);
        let spec = command_utils::add_verbose_flags(spec);
        let spec = command_utils::add_from_flag(spec);
        let spec = command_utils::add_path_flag(spec);
        let spec = command_utils::add_wait_for_recheck_flag(spec);
        spec.flag(
            "--in-place",
            &arg_spec::truthy(),
            "Overwrite the input file or file specified by the path flag",
            None,
        )
        .flag(
            "--force",
            &arg_spec::truthy(),
            "Write the results even if errors are encountered",
            None,
        )
        .anon("file", &arg_spec::required(None, arg_spec::string()))
    }

    fn handle_error(code: flow_common_exit_status::FlowExitStatus, msg: &str) -> ! {
        eprintln!("{}", msg);
        flow_common_exit_status::exit(code);
    }

    fn select_output_channel(
        in_place: bool,
        path: Option<&str>,
        source_path: &str,
    ) -> Box<dyn Write> {
        match (in_place, path) {
            (false, _) => Box::new(std::io::stdout()),
            (true, Some(p)) => Box::new(std::fs::File::create(p).unwrap_or_else(|_| {
                handle_error(
                    flow_common_exit_status::FlowExitStatus::PathIsNotAFile,
                    &format!("failed to open output file: {}", p),
                )
            })),
            (true, None) => Box::new(std::fs::File::create(source_path).unwrap_or_else(|_| {
                handle_error(
                    flow_common_exit_status::FlowExitStatus::PathIsNotAFile,
                    &format!("failed to open output file: {}", source_path),
                )
            })),
        }
    }

    #[allow(dead_code)]
    const AVG_ERROR_SIZE: usize = 100;

    fn append_errors(errors: &[String]) -> String {
        let mut buff = String::with_capacity(AVG_ERROR_SIZE * errors.len());
        for error in errors {
            buff.push_str(error);
        }
        buff
    }

    fn handle_ok(
        patch: &replacement_printer::Patch,
        errors: &[String],
        input: &FileInput,
        in_place: bool,
        forced: bool,
        path: Option<&str>,
        source_path: &str,
    ) {
        let write_patch = |content: &str| {
            let mut out = select_output_channel(in_place, path, source_path);
            out.write_all(replacement_printer::print(patch, content).as_bytes())
                .unwrap();
            out.flush().unwrap();
        };
        match (input.content_of_file_input(), errors, forced) {
            (Ok(content), [], _) | (Ok(content), _, true) => {
                eprint!("{}", append_errors(errors));
                write_patch(&content);
            }
            (Ok(_), errors, false) => handle_error(
                flow_common_exit_status::FlowExitStatus::UnknownError,
                &append_errors(errors),
            ),
            (Err(msg), _, _) => {
                handle_error(flow_common_exit_status::FlowExitStatus::UnknownError, &msg)
            }
        }
    }

    fn main(args: &arg_spec::Values) {
        let base_flags = command_utils::get_base_flags(args);
        let (connect_flags, json, _pretty) = command_utils::get_connect_and_json_flags(args);
        let root_arg =
            command_spec::get(args, "--root", &arg_spec::optional(arg_spec::string())).unwrap();
        let _strip_root_arg = command_spec::get(args, "--strip-root", &arg_spec::truthy()).unwrap();
        let verbose = command_utils::verbose_flags(args);
        let path =
            command_spec::get(args, "--path", &arg_spec::optional(arg_spec::string())).unwrap();
        let wait_for_recheck = command_spec::get(
            args,
            "--wait-for-recheck",
            &arg_spec::optional(arg_spec::bool_flag()),
        )
        .unwrap();
        let in_place = command_spec::get(args, "--in-place", &arg_spec::truthy()).unwrap();
        let forced = command_spec::get(args, "--force", &arg_spec::truthy()).unwrap();
        let file =
            command_spec::get(args, "file", &arg_spec::required(None, arg_spec::string())).unwrap();

        let source_path = command_utils::expand_path(&file);
        let input = command_utils::get_file_from_filename_or_stdin(
            &spec().name,
            Some(source_path.as_str()),
            path.as_deref(),
        );
        let root = command_utils::get_the_root(
            &base_flags.flowconfig_name,
            root_arg.as_deref(),
            Some(&input),
        );
        let flowconfig_name = &base_flags.flowconfig_name;
        if !json && verbose.is_some() {
            eprintln!("NOTE: --verbose writes to the server log file");
        }
        let request = server_prot::request::Command::AUTOFIX_EXPORTS {
            input: input.clone(),
            verbose,
            wait_for_recheck,
        };
        let result = command_utils::connect_and_make_request(
            flowconfig_name,
            &connect_flags,
            &root,
            &request,
        );
        match result {
            server_prot::response::Response::AUTOFIX_EXPORTS(Err(err)) => {
                handle_error(flow_common_exit_status::FlowExitStatus::UnknownError, &err)
            }
            server_prot::response::Response::AUTOFIX_EXPORTS(Ok((patch, errors))) => handle_ok(
                &patch,
                &errors,
                &input,
                in_place,
                forced,
                path.as_deref(),
                &source_path,
            ),
            _ => handle_error(
                flow_common_exit_status::FlowExitStatus::UnknownError,
                "Flow: invalid server response",
            ),
        }
    }

    pub(super) fn command() -> command_spec::Command {
        command_spec::command(spec(), main)
    }
}

pub(crate) fn command() -> command_spec::Command {
    let main = |args: &arg_spec::Values| {
        let (cmd, argv) = command_spec::get(
            args,
            "subcommand",
            &arg_spec::required(
                None,
                arg_spec::command_flag(vec![
                    ("insert-type", "insert-type"),
                    ("missing-local-annot", "missing-local-annot"),
                    ("exports", "exports"),
                ]),
            ),
        )
        .unwrap();
        let cmd = match cmd {
            "insert-type" => insert_type::command(),
            "missing-local-annot" => missing_local_annot::command(),
            "exports" => exports::command(),
            _ => unreachable!(),
        };
        command_utils::run_command(&cmd, &argv);
    };
    let spec = command_utils::subcommand_spec(
        "autofix",
        "Automatically insert type annotations",
        command_spec::Visibility::Internal,
        vec![
            (
                "insert-type",
                "insert-type",
                insert_type::command().doc().to_owned(),
            ),
            (
                "missing-local-annot",
                "missing-local-annot",
                missing_local_annot::command().doc().to_owned(),
            ),
            ("exports", "exports", exports::command().doc().to_owned()),
        ],
    );
    command_spec::command(spec, main)
}
